### ISSUE TITLE
Buffer overflows with strings as global variables

### ISSUE THREAD

**Issue by simon6433c** [michael-lehn/abc-llvm#2]
#### Description
It is possible to create buffer overflows with global string variables.
For convenience I will call them strings, although they are actually pointers to ```u8``` or arrays of ```u8```.

#### To Reproduce
The following code should illustrate the problem. 

``` abc
@ <stdio.hdr>
extern fn exit (:int);

global string0 : array [32] of u8 = "Malicious Buffer Overflow       "; // Misses the zero char at the end
global hexu64 : u64 = 0x6161616161616161; // 8 printable bytes buffered to get to the next string address
global hexu64_ : u64 = 0x6161616161616161; // 8 printable bytes buffered to get to the next string address
global hello1 : array [6] of u8 = "Hallo1"; // Misses the zero char at the end
global hello2 : array [7] of u8 = "Hallo2"; // Has the zero char
global hello3 : array [6] of u8 = "Hallo3"; // Misses the zero char at the end
global hello4 : array [6] of u8 = "Hallo4"; // Misses the zero char at the end

fn main () {
    printf (">> %zu @ %s\n", string0, string0);
    printf (">> %zu @ %s\n", hello1, hello1);
    printf (">> %zu @ %s\n", hello2, hello2);
    printf (">> %zu @ %s\n", hello3, hello3);
    printf (">> %zu @ %s\n", hello4, hello4);
    printf (">> %zu @ %s\n", &hexu64, &hexu64);
    exit (0);
}
```

The output of this snippet on my machine is the following :
``` plain
>> 140356855304208 @ Malicious Buffer Overflow       aaaaaaaaaaaaaaaaHallo1Hallo2
>> 140356855304256 @ Hallo1Hallo2
>> 140356855304262 @ Hallo2
>> 140356855304269 @ Hallo3Hallo4
>> 140356855304275 @ Hallo4
>> 140356855304240 @ aaaaaaaaaaaaaaaaHallo1Hallo2
```

It illustrates, that it is possible to have strings stored in the global namespace "escape" its bounds because the array of  ```u8``` is  allowed to be one byte to small to store the  terminating zero-byte. 

In the example ```global hello2 : array [7] of u8 = "Hallo2"``` it is shown, that this overflow does not happen, when the array is one byte bigger, so that it can now contain the zero-byte.

The example print of ```string0``` even shows, that it is possible to overflow not only the to the neighboring string,  but even read through other variables, that should normally not be printed as text. 

#### Reason and possible Solutions
Looking at the assembler code the behavior can be explained.
``` asm
string0:
	.ascii	"Malicious Buffer Overflow       "
	.size	string0, 32

	.type	hexu64,@object
	.p2align	3, 0x0
hexu64:
	.quad	7016996765293437281
	.size	hexu64, 8

	.type	hexu64_,@object
	.p2align	3, 0x0
hexu64_:
	.quad	7016996765293437281
	.size	hexu64_, 8

	.type	hello1,@object
hello1:
	.ascii	"Hallo1"
	.size	hello1, 6

	.type	hello2,@object
hello2:
	.asciz	"Hallo2"
	.size	hello2, 7

	.type	hello3,@object
hello3:
	.ascii	"Hallo3"
	.size	hello3, 6

	.type	hello4,@object
hello4:
	.ascii	"Hallo4"
	.size	hello4, 6

	.type	.L0,@object
	.section	.rodata,"a",@progbits
```
This snippet only contains the global space for variables. The assembler code for the main function was ommitted.

As the assembler code shows, only the ```hello2``` variable is stored as an ```.asciiz```, which denotes a zero terminated ascii string.
The other strings are only stored as ```.ascii```, which means that they are not zero terminated. 
This may be intended behavior, since if arrays of ```u8``` are used, the size may be known. 

The issue is, that it is required to use arrays to store global strings and those arrays to not require, that the last entry is a zero terminator.

Possible solutions are either to force the array size of the  ```u8``` array to be big enough to store the zero terminator, if string literals are stored into them. 
This may be a easy solution, since a length check is already in place as shown below. But it only checks if the array is long enough to store the string excluding the zero terminator.
``` abc
global hello5 : array [5] of u8 = "Hallo5";
```
``` plain
global hello5 : array [5] of u8 = "Hallo5";
                                  ^^^^^^^^^
overflow.abc:13.35-13.43: : error: : excess elements in array initializer
```

An other solution would be to allow the storing of zero terminated strings in  ```u8``` pointers like in the following example. 
This would also simplify writing global string variables, since you do not need to update the length of the global array all of the time, just to have enough space for the string or to prevent global space to be allocated without ever being used.
``` abc
global moin : -> u8 = "Moin";
```
``` plain
gen::cat: can not cast 'array [5] of u8' to '-> u8'
abc: gen/cast.cpp:39: llvm::Value* gen::cast(gen::Value, const abc::Type*, const abc::Type*): Assertion `0' failed.
```
As shown this is currently not possible and it even creates an assertion. 
This is something that also should be fixed and may be content of a future issue, when I understand it good enough. 

## TL;DR
It is possible to create overflows with string variables, since the compiler presumably thinks the length of a string literal is only the written content without the zero terminator (i.e. the empty string ```""``` has length 0 but it should actually contains 1 character: the zero terminator).

**Comment by michael-lehn** [comment]
Confirmed. This behavior was actually intentional: When used for initialization a string literals like `"hello"` is treated as `{'h', 'e', 'l', 'l', 'o'}`. That means

-  This
   ```
   global hello6 : array [6] of u8 = "hello";
   ```
   is equivalent to
   ```
   global hello6 : array [6] of u8 = {'h', 'e', 'l', 'l', 'o'};
   ```
   so the first five element of the array a initialized with characters, the rest wit zero bytes.
-  This
   ```
   global hello4 : array [4] of u8 = "hello";
   ```
   is equivalent to
   ```
   global hello6 : array [4] of u8 = {'h', 'e', 'l', 'l', 'o'};
   ```
   and gives an compile time error (excess elements in array initializer)
-  This
   ```
   global hello5 : array [5] of u8 = "hello";
   ```
   is equivalent to
   ```
   global hello6 : array [5] of u8 = {'h', 'e', 'l', 'l', 'o'};
   ```
   and initializes all elements. And yes, that means the array is not zero terminated.


But I agree, string as special kind of arrays. The simplest way to fix this should be to treat `"hello"` is treated as `{'h', 'e', 'l', 'l', 'o', 0}`, i. e. as an initializer list with an extra zero byte.

I post back after committing this change
